### PR TITLE
B63telemetry probes (re #62 and #63) WIP

### PR DESCRIPTION
**Pull request by raymak**

### CODE REVIEW

File: addon/lib/Feature.jsm [hunk 1]
```diff
@@ -695,9 +726,11 @@ class Feature {
       (panelHiddenTime - this.panelShownTime) / 1000;
     this.log.debug(`${panelType} was open for ${Math.round(panelOpenTime)} seconds.`);
     this.telemetry({
-      event: `${panelType}-hidden`,
-      secondsPanelWasShowing: Math.round(panelOpenTime).toString(),
-    });
+        message_type: "event",
+        event: "panel-hidden",
+        panel_type: panelType,
+        secondsPanelWasShowing: Math.round(panelOpenTime).toString(),
```
Comment (biancadanforth) [inline addon/lib/Feature.jsm]: This is actually currently in ms, not s. I can change that easily.

File: addon/content/new-tab-variation.js [hunk 2]
```diff
@@ -21,6 +21,16 @@ class TrackingProtectionStudy {
...
     addMessageListener("TrackingStudy:Uninstalling", this);
+
+    this.initTimer();
+  }
+
+  initTimer() {
+    this.openingTime = Math.floor(Date.now() / 1000);
+    this.contentWindow.addEventListener("beforeunload", ()=> {
+      sendAsyncMessage("TrackingStudy::NewTabOpenTime",
+        Math.floor(Date.now() / 1000) - this.openingTime);
```
Comment (biancadanforth) [inline addon/content/new-tab-variation.js]: This method produces wildly inaccurate times for the newtab page being open! I have no idea why as the math seems fine:
wildlyinaccuratetimes

Comment (biancadanforth) [reply addon/content/new-tab-variation.js]: Also, there should be a corresponding `removeEventListener` for this as well.

File: addon/lib/Feature.jsm (hunk 1 above)
Comment (raymak) [reply addon/lib/Feature.jsm]: I'll change it.

File: addon/content/new-tab-variation.js (hunk 2 above)
Comment (raymak) [reply addon/content/new-tab-variation.js]: I will add the removeEvenetListener

File: addon/lib/Feature.jsm [hunk 3]
```diff
@@ -954,6 +989,12 @@ class Feature {
       // if "fast" treatment, convert counter from ms to seconds and add unit "s"
       const label = this.treatment === "private" ? counter
         : `${Math.ceil(counter / 1000)}s`;
+
+      this.telemetry({
```
[the remark this answers is not in the document]
Comment (raymak) [reply addon/lib/Feature.jsm]: I like your idea!

File: addon/lib/Feature.jsm (hunk 1 above)
Comment (raymak) [reply addon/lib/Feature.jsm]: Oh are you sure about this? (panelHiddenTime - this.panelShownTime) / 1000 is seconds?

Comment (biancadanforth) [reply addon/lib/Feature.jsm]: Oh maybe not... hmm... what did I see that was in ms? Ah well. Please disregard then.

File: addon/content/new-tab-variation.js (hunk 2 above)
Comment (biancadanforth) [reply addon/content/new-tab-variation.js]: I figured out why the timer is broken! Thanks to Mardak of the Activity Stream team.

**Tl;dr: Set the pref `"browser.newtab.preload"` to `false` at the beginning of the study and reset it at the end of the study. Kamyar, can you add this change to your list of action items?**

_A longer explanation:_
I thought perhaps there was a problem with `Date.now()` and I tried to use a relative measure of time instead using `setInterval` and incrementing a counter every second. What I discovered was that the frame script's `init` method (where the timer is initialized) is actually being called twice the first time I visit `about:newtab`.

Per Mardak:
>  preloading is only for getting the next about:newtab ready. the very first new tab is not preloaded though
the first about:newtab is not preloaded, opening a new tab will trigger a regular load and a preload for the next one
when you open a second about:newtab, the preloaded one was already loaded and it now appears. then a new preload is triggered for the 3rd time
there is a perf impact as without preloading, the browser spends time loading javascript, react rendering, waiting for messages, etc

Sure enough, when I tried this out by setting that pref in this branch, the timer works!

File: addon/lib/Feature.jsm [hunk 4]
```diff
@@ -1051,6 +1063,9 @@ class Feature {
   }
 
   async endStudy(reason, shouldResetTP = true) {
+    await this.reportBehaviorSummary();
```
Comment (biancadanforth) [inline addon/lib/Feature.jsm]: Good idea!

**Comment by biancadanforth** [comment]
Just documenting the things we discussed yesterday afternoon about what's left for this PR:
- Average and median pageAction counter values when button is clicked.
- Covariate of the top 2-3 ad blocker webextensions
- Removing the call to initBehaviorSummary in constructor
- Write TELEMETRY.md

File: addon/lib/Feature.jsm [hunk 5]
```diff
@@ -163,12 +168,17 @@ class Feature {
...
     // since we are initializing built-in TP based on the treatment.
     this.addBuiltInTrackingProtectionListeners();
+
+    await this.initBehaviorSummary();
   }
 
   addContentMessageListeners() {
     // content listener
     Services.mm.addMessageListener("TrackingStudy:InitialContent", this);
+    Services.mm.addMessageListener("TrackingStudy::NewTabOpenTime", this);
```
Comment (pdehaan) [inline addon/lib/Feature.jsm]: Why does one have `TrackingStudy:*` and the other has `TrackingStudy::*` (one-vs-two `::`)?

File: addon/lib/Feature.jsm [hunk 6]
```diff
@@ -932,6 +1009,25 @@ class Feature {
           this.introPanelMessages[this.treatment],
           isIntroPanel
         );
+
+        let counter = this.treatment === "private" ?
+        this.state.blockedResources.get(win.gBrowser.selectedBrowser) :
+        this.state.timeSaved.get(win.gBrowser.selectedBrowser);
+
+        // log page action click
+        Storage.get("behavior-summary").then((behaviorSummary) => {
```
Comment (pdehaan) [inline addon/lib/Feature.jsm]: Not sure if this should be `async` + `await`ed to be consistent with other usages, or if not worth the effort.

Comment (biancadanforth) [reply addon/lib/Feature.jsm]: Not a blocker; but I agree consistency is a lofty ideal to aspire to.

File: addon/lib/Feature.jsm (hunk 5 above)
Comment (biancadanforth) [reply addon/lib/Feature.jsm]: That's how you know which listeners Kamyar added and which listeners I added. :)

Comment (pdehaan) [reply addon/lib/Feature.jsm]: in that case, can I propose something like single colons and we append your names to the events? Something like `"TrackingStudy:IntialContent:Bianca"` and `"TrackingStudy:NewTabOpen:Kamyar"`?

File: addon/lib/Feature.jsm [hunk 7]
```diff
@@ -214,6 +242,52 @@ class Feature {
...
+    let containsAddon = false;
+    AddonManager.getAllAddons().then((addons) => {
+      for (let addon of addons) {
+        if (ADBLOCKER_ID_LIST.includes(addon.id)) containsAddon = true;
+      }
+    })
+
+    // co-variates
+    const oldestTimestamp = await ((new ProfileAge()).getOldestProfileTimestamp())
+    const ProfileAgeDays = Math.round((Date.now() - oldestTimestamp) / (1000 * 3600 * 24));
```
Comment (pdehaan) [inline addon/lib/Feature.jsm]: nit: Should this be lowercase `profileAgeDays`?

File: addon/lib/Feature.jsm [hunk 8]
```diff
@@ -214,6 +242,52 @@ class Feature {
...
+      intro_accept: "false",
+      intro_reject: "false",
+      badge_clicks: String(0),
+      panel_open_times: String([]),
+      panel_open_times_median: String(0),
+      panel_open_times_mean: String(0),
+      new_tab_open_times: String([]),
+      new_tab_open_times_median: String(0),
+      new_tab_open_times_mean: String(0),
+      page_action_counter: String([]),
```
Comment (pdehaan) [inline addon/lib/Feature.jsm]: Sorry, I'm not too knowledgeable on Storage vs `String()` constructors, but...
IIUC, calling `String([])` will return an empty string, `""`. Whereas `JSON.stringify([])` will return `"[]"`. I have no idea if any of that matters, I guess it all depends on who consumes the data at the other end (or how we retrieve and rehydrate it from storage):

```js
String([]); // ""
String(["one fish", "two fish"]) //"one fish,two fish"

JSON.stringify([]) // "[]"
JSON.stringify(["red fish", "blue fish"]) // "[\"red fish\",\"blue fish\"]"
```

File: addon/lib/Feature.jsm (hunk 6 above)
Comment (pdehaan) [reply addon/lib/Feature.jsm]: I did a quick Ctrl+F on the PR files and found 4 `.then(` strings, and 0 `.catch(` strings. I can individually comment on each line if that in some way helps in any possible efforts to `await` our `async`s. 🤷‍♀️ 

Otherwise, I'm fine with merging without those changes and I can file a separate bug to async-ify any lurking promises.

**Comment by pdehaan** [comment]
Which issues will this PR fix?
I think I only see references to `(fixes #63, fixes #87)` in a commit message, both of which should get auto-closed when that commit gets merged.
Wasn't sure if this addressed anything else, and we should update the initial GitHub commit message/comment above, so everything gets linked and closed automatically.

File: addon/lib/Feature.jsm (hunk 6 above)
Comment (biancadanforth) [reply addon/lib/Feature.jsm]: Let's file a separate bug for that if you don't mind!

File: addon/lib/Feature.jsm [hunk 9]
```diff
@@ -936,6 +1095,27 @@ class Feature {
           this.introPanelMessages[this.treatment],
           isIntroPanel
         );
+
+        let counter = this.treatment === "private" ?
+        this.state.blockedResources.get(win.gBrowser.selectedBrowser) :
+        Math.round(this.state.timeSaved.get(win.gBrowser.selectedBrowser) / 1000);
```
Comment (biancadanforth) [inline addon/lib/Feature.jsm]: For historical reasons, can you change `Math.round` to `Math.ceil`? I use `Math.ceil` to calculate the pageAction badge value (just rounding up so that fractions of a second show as 1s, which looks more effective than 0s). That way, there won't be a rounding error between the value recorded in the summary ping and the actual seen value on the pageAction button.

File: addon/lib/Feature.jsm [hunk 10]
```diff
@@ -971,6 +1151,9 @@ class Feature {
   }
 
   async endStudy(reason, shouldResetTP = true) {
+    // returning the Activity Stream preloading to its default state
+    Services.prefs.clearUserPref("browser.newtab.preload");
```
Comment (biancadanforth) [inline addon/lib/Feature.jsm]: Can you actually put this in the `uninit` method? Otherwise this pref doesn't get cleared in the ending condition of `expired` from `bootstrap.js`.
I actually create a new instance of `Feature` on every Firefox startup and `uninit` it on every shutdown.

The only reason I have a `Feature.endStudy` was to know whether or not I should reset TP, because in some cases, the study has ended because the user explicitly changed their TP setting from their assigned setting based on the treatment. In those cases, I don't want to reset TP at the end of the study.

File: addon/lib/Feature.jsm [hunk 11]
```diff
@@ -979,6 +1162,7 @@ class Feature {
   }
 
   async uninit() {
+    await this.reportBehaviorSummary();
```
Comment (biancadanforth) [inline addon/lib/Feature.jsm]: Since I call `uninit` on Bootstrap shutdown, this will get called anytime Firefox shuts down, not just at the very end... Consider calling `this.feature.reportBehaviorSummary()` in bootstrap?

```javascript
// in bootstrap.shutdown()...
if (isUninstall) {
  // other stuff
  await this.feature.reportBehaviorSummary();
  await Storage.clear();  // will have to import Storage.jsm into bootstrap.js
}
```

**Comment by biancadanforth** [comment]
I claim this PR will fix #87 , #63 and #31 . Once TELEMETRY.md is written (in a separate PR), that combined with this PR will close #62 .

File: addon/lib/Feature.jsm (hunk 11 above)
Comment (raymak) [reply addon/lib/Feature.jsm]: I intentionally added it to uninit() to make sure it's still reported even if telemetry fails on shutdown due to some race condition or another issue. Your call though. From an analysis standpoint, it's easier to have only one ping from each user, but it's not that difficult to only find the latest ping either.

File: addon/lib/Feature.jsm (hunk 8 above)
Comment (raymak) [reply addon/lib/Feature.jsm]: I agree that JSON.stringify is a better choice.

File: addon/lib/Feature.jsm (hunk 9 above)
Comment (raymak) [reply addon/lib/Feature.jsm]: For consistency, I changed them all to `Math.ceil`.

File: addon/lib/Feature.jsm (hunk 8 above)
Comment (pdehaan) [reply addon/lib/Feature.jsm]: I don't know about better or worse (or write our own stringifier, if we think we'd use it in multiple places). But I do know calling `String(true)`, `String([])`, and `String(0)` gives different results than `JSON.stringify(true)`, `JSON.stringify([])`, and `JSON.stringify(0)` so we'd need to make sure our Telemetry data (or whatever we're sending this packet to) is still valid at the other end.

File: addon/lib/Feature.jsm (hunk 11 above)
Comment (biancadanforth) [reply addon/lib/Feature.jsm]: Okay thanks for that insight! I tested it in a bunch of different uninstall cases, and it seems to get sent properly. I'll leave it this way for now.
